snapshots/overlay/overlay.go

```diff
@@ -102,22 +130,92 @@ func NewSnapshotter(root string, opts ...Opt) (snapshots.Snapshotter, error) {
 		return nil, err
 	}
 
-	if err := os.Mkdir(filepath.Join(root, "snapshots"), 0700); err != nil && !os.IsExist(err) {
-		return nil, err
+	firstSnapshot := true
+	if err := os.Mkdir(filepath.Join(root, "snapshots"), 0700); err != nil {
+		if !os.IsExist(err) {
+			return nil, err
+		}
+		firstSnapshot = false
```

Comment:
I'm not sure if this can be removed too. IIUC this is only used by [this][]. But we can just check for that and either disable it or return an error.


[this]: https://github.com/containerd/containerd/pull/7141/files#diff-d0387518abb075451d8d197f968976633e14ad7a8f6fbc665b2906afa7998e5eR196-R201

Reply:
Could, you, pls, clarify your point here? `firstSnapshot` is needed to figure out whether we already have snapshots with `metacopy=on` or not. You mean to just check for existence directory `<root>/snapshots`?

Reply:
Nope, IIUC it is only used for what I linked before: https://github.com/containerd/containerd/pull/7141/files#diff-d0387518abb075451d8d197f968976633e14ad7a8f6fbc665b2906afa7998e5eR196-R201

I don't think we need to know if it is the first snapshot to do that if and log/return an error. Do we?

Reply:
If `metacopy=on` by kernel and this is the first snapshot we can use `metacopy=off` by config, if we already have snapshots we can't. If we remove this check we won't be able to use `metacopy=off`  for first snapshot if `metacopy=on` by kernel, though it is allowed -- `if metacopyCtrl=true && isMetacopyOn=true && (!(config.Metacopy=false))=true` -> `metacopyOn = true`

Reply:
Why wouldn't you be allowed to use metacopy=off if it is on by the kernel and you don't know if it is the first snapshot?

I mean, if you don't know if it is the first snapshot or not, and there is no file in the root signaling if metacopy should be used or not, you can honor the config and hope for the best. You don't have to treat the first snapshot special in any way for that.

Maybe it will be nice to know if it is the first snapshot, for when you create the file signaling metacopy or not. In the first snapshot you can be confident when you create the file (although don't we have to look at the parent snapshots to know if metacopy is needed or not?).

But don't we want to create the file even if it is not the first snapshot? If metacopy is on, not the first snapshot and we create the snapshotter, we better use metacopy on the next runs too. Right? So, I'm not sure we need this firstSnapshot thing

Reply:
>Why wouldn't you be allowed to use metacopy=off if it is on by the kernel and you don't know if it is the first snapshot?

I meant if we already have snapshots with `metacopy=on`.

>I mean, if you don't know if it is the first snapshot or not, and there is no file in the root signaling if metacopy should be used or not, you can honor the config and hope for the best. You don't have to treat the first snapshot special in any way for that.

Actually, as I wrote above the only use case with this `firstSnapshot` is to use `metacopy=off` by config for first snapshot if kernel says `metacopy=on`. If we check only kernel for that -- it says `metacopy=on`, so we assume that we already have snapshots with `metacopy=on`, so it is not allowed to create a new one with `metacopy=off`, but if this is the first snapshot we can accept any of `metacopy=off/on` options, this is the use case.

Life gets easier if we can rely on user's input (I mean metacopy value from config), but I'm really not sure in this due to it brings not so much overhead, but prevents user from creating a "broken" snapshot.   

>although don't we have to look at the parent snapshots to know if metacopy is needed or not?

We can check for metacopy from parents. Firstly, I looked exactly at this variant, but I don't think it is really needed it only brings overhead:

1. We can't do this check for snapshots which have been created before this change, we can only initialize  metacopy value for parents according to the kernel value, but it is the same as we do now (I mean we check for kernel option).

2. If we have snapshots that created only with this change --  logic's getting a little bit complicated at the first glance, we need to combine metacopy options from several parents, these options may not be the same. There was another case with this related to exactly different options from parents, but I can't recollect now :smiling_face_with_tear: , sorry ^_^.

>But don't we want to create the file even if it is not the first snapshot? 

In this version, file's created regardless of `firstSnapshot`, only if it doesn't exist.

Reply:
> Actually, as I wrote above the only use case with this `firstSnapshot` is to use `metacopy=off` by config for first snapshot if kernel says `metacopy=on`. If we check only kernel for that -- it says `metacopy=on`, so we assume that we already have snapshots with `metacopy=on`, so it is not allowed to create a new one with `metacopy=off`, but if this is the first snapshot we can accept any of `metacopy=off/on` options, this is the use case.

I don't see the point, but if you do, okay.

Having said that, I think that if we don't have the file signaling if metacopy should be on/off for the snapshot, I think we should honor the config. It doesn't matter if it is the first snapshot or not.



> 2\. If we have snapshots that created only with this change 

It doesn't need to be all snapshots created only after this change. Just some.

> --  logic's getting a little bit complicated at the first glance, we need to combine metacopy options from several parents,

Yes, we need an "or" from the metacopy param of all parents in that case. If some had it enabled, then it needs to be enabled.

> these options may not be the same. There was another case with this related to exactly different options from parents, but I can't recollect now smiling_face_with_tear , sorry ^_^.

The question is, can this work correctly without handling the parents metacopy? If it can, then sure, let's make it simpler and just ignore it.

Reply:
>I don't see the point, but if you do, okay.

Once I update PR it should look clearer. If it will turn out that it's not needed, sure, I'll remove it.

>The question is, can this work correctly without handling the parents metacopy? If it can, then sure, let's make it simpler and just ignore it.

Sure, it can, we just have one global value instead of a local one for every snapshot. 

I would say that now there's no way to completely prevent the original issue, even if we'll have metacopy options from parents, for those "some" snapshots, which have been created before this change we have to rely on the kernel metacopy option and assume that their parents used the same option as well, but a user could switch `/sys/module/overlay/parameters/metacopy` by himself/herself  a few times earlier and may already have "broken" snapshots. For me, to have metacopy options for parents looks like an overkill.

I'm going to update PR on days or may be next week according to all these comments and remove unnecessary code (I have a draft version), so let's go back to this issue once PR's updated, I think it'd be better, because we'll have a clearer view on this.

If we'll come to a conclusion that parent values of metacopy are really needed, sure, I'll add them.

Reply:
Sorry, forgot to send the message before: SGTM! :)